File: _history/server/client_code/20260314t1600/_iworker.py
from ..console import console, log
from ..document import document
from ..meta import meta
from ..native import native
from ..tools import Promise
from ..use import use
from ..window import window
from ..works import works




_is = use("@/rollo/")["is"]


def iframe(**data):
    """Sets props on iframe."""
    # NOTE Intended for setting attrs, CSS vars and classes
    promise = Promise()
    channel = native.MessageChannel()

    def onmessage(event):
        channel.port1.close()
        promise(event.data)

    channel.port1.onmessage = onmessage
    window.parent.postMessage(
        dict(
            type="iframe",
            data=data,
        ),
        meta.client.origin,
        [channel.port2],
    )
    result = promise.wait()
    # print("Resize confirmation:", result)  ##
    return result


def on_resize_y(event):
    """Sets height on iframe."""
    iframe(__height=f'{event.detail}px')


class iworker(works.HtmlTemplate):
    def __init__(
        self,
        page: str = None,
        path: str = None,
        query: dict = None,
        session: str = None,
        targets: list = None,
        **props,
    ):
        if not meta.iworker:
            return

        self._ = {}

        # Get server data
        targets = tuple(targets.sort() or targets) if targets else tuple()
        self(session=session, targets=targets)
        use.assets.meta.server.targets = targets
        # log("session:", session, trace=__file__)  ##
        self.connect()

    def __call__(self, **updates):
        self._.update(updates)
        return self

    def __getitem__(self, key: str):
        return self._.get(key)

    def __getattr__(self, key: str):
        return self[key]

    def connect(self):
        """Performs handshake and sets up request handler."""
        # Guard against redundant calls
        if self.connected:
            return
        self(connected=True)

        @window.on()
        def message(event):
            """Handles request-response."""
            if not self.qualify(event, type="request"):
                return
            # Extract data from event
            port = event.ports[0]
            specifier = event.data.specifier
            args = getattr(event.data, "args", [])
            kwargs = getattr(event.data, "kwargs", {})
            submission = getattr(event.data, "submission", None)
            visible = getattr(event.data, "visible", False)


            if visible:
                use.app.addEventListener('_resize_y', on_resize_y)



            # Get target
            target = use(specifier, test=getattr(event.data, "test", False))
            # Call target and post result
            try:
                if isinstance(target, type):
                    is_component = issubclass(target, works.HtmlTemplate)
                    if "__init__" in target.__dict__:
                        instance = target(submission=submission, visible=visible)
                    else:
                        instance = target()
                    if is_component:
                        use.main.child = instance
                else:
                    instance = target

                result = instance(*args, **kwargs) if callable(instance) else None
                if isinstance(result, Promise):
                    result = result.wait()

                port.postMessage(dict(result=result, submission=submission))
            except Exception as error:
                port.postMessage(
                    dict(error=f"Error when calling '{specifier}': {str(error)}")
                )

            if visible:
                use.app.removeEventListener('_resize_y', on_resize_y)
                iframe(__height=0)

        # Handshake
        promise = Promise()
        channel = native.MessageChannel()

        def onmessage(event):
            log("Init data from client:", event.data)  ##
            channel.port1.close()
            promise(event.data)

        channel.port1.onmessage = onmessage
        # Send init data to client
        window.parent.postMessage(
            dict(
                type="ready",
                detail=dict(
                    packages=use.packages.keys(),
                    server=dict(session=self.session, targets=self.targets),
                ),
            ),
            meta.client.origin,
            [channel.port2],
        )
        promise.wait()

    @staticmethod
    def qualify(event, **kwargs):
        if event.origin != meta.client.origin:
            return
        if not _is.object(event.data):
            return
        if "type" in kwargs and event.data.type != kwargs["type"]:
            return
        return True

File: client/public/parcels/iworker/iworker.js
const { component: b } = await use("@/rollo/"), { frame: q } = await use("@/frame/"), o = b.iframe({
  id: "iworker",
  name: "iworker",
  src: `${use.meta.server.origin}/iworker?origin=${location.origin}`,
  slot: "iworker"
  //__height: 0,
});
q.append(o);
class u {
  static create = (...e) => new u(...e);
  #e = {
    channel: new MessageChannel()
  };
  constructor() {
  }
  close() {
    this.#e.channel.port1.close();
  }
  receive(e) {
    this.#e.channel.port1.onmessage = e;
  }
  send(e = {}) {
    o.contentWindow.postMessage(e, use.meta.server.origin, [
      this.#e.channel.port2
    ]);
  }
}
const { is: E } = await use("@/rollo/");
class d {
  static create = (...e) => new d(...e);
  #e = {};
  constructor(e, { type: s } = {}) {
    this.#e = {
      event: e,
      ok: e.origin === use.meta.server.origin && E.object(e.data) && (!s || e.data.type === s),
      type: s
    };
  }
  get detail() {
    return this.#e.event.data.detail || null;
  }
  get event() {
    return this.#e.event;
  }
  get ok() {
    return this.#e.ok;
  }
  get submission() {
    return this.#e.event.data.submission || null;
  }
  get type() {
    return this.#e.type || null;
  }
  respond(e = {}) {
    this.#e.event.ports[0].postMessage({
      detail: e,
      submission: this.submission,
      type: this.type
    });
  }
}
const v = /* @__PURE__ */ (() => {
  let r = 0;
  return () => r++;
})();
class m {
  static create = (...e) => new m(...e);
  #e = {};
  constructor() {
    this.#e.onmessage = (e) => {
      const s = d.create(e, { type: "iframe" });
      if (!s.ok)
        return;
      const t = s.detail || {};
      use.meta.DEV && //console.log("Client received updates for iframe:", updates); ////
      o.update(t), s.respond(t);
    };
  }
  get active() {
    return this.#e.active;
  }
  start() {
    this.active || (this.#e.active = !0, window.addEventListener("message", this.#e.onmessage));
  }
  stop() {
    this.active && (delete this.#e.active, window.removeEventListener("message", this.#e.onmessage));
  }
}
const { is: f } = await use("@/rollo/"), y = (r) => {
  const e = r.find((t, n) => !n && f.object(t)) || {}, s = r.find((t, n) => (!n || n === 1) && f.object(t) && t !== e) || {};
  return r = r.filter((t, n) => t !== e && t !== s), [e, s, r];
}, l = new class {
  #e = {
    /* Tail of the show() chain. Initialized as an
    already-resolved promise so the very first caller starts immediately
    without waiting. Each call advances this pointer to its own promise,
    forming a linked chain that enforces one-at-a-time execution. */
    queue: Promise.resolve(),
    sync: m.create()
  };
  constructor() {
  }
  get sync() {
    return this.#e.sync;
  }
  async request(r, ...e) {
    const [s, t, n] = y(e);
    return await new Promise((p, i) => {
      const a = u.create();
      a.receive((c) => {
        c.data.error ? i(c.data.error) : p(c.data.result), a.close();
      }), a.send({
        submission: v(),
        type: "request",
        specifier: r,
        args: n,
        kwargs: t,
        options: s
      });
    });
  }
  async show(r, ...e) {
    const s = async () => {
      const [n, g, p] = y(e), { visible: i = !0 } = n, a = v();
      return new Promise((c, k) => {
        const h = u.create();
        i === "popover" ? o.update({ popover: "manual", __height: "100vh" }).showPopover() : this.sync.start(), h.receive((w) => {
          w.data.error ? k(w.data.error) : c(w.data.result), h.close(), o.update({ __height: 0 }), i === "popover" ? o.update({ popover: null }) : this.sync.stop();
        }), h.send({
          submission: a,
          type: "show",
          specifier: r,
          args: p,
          kwargs: g,
          options: n,
          visible: i
        });
      });
    }, t = this.#e.queue.then(s);
    return this.#e.queue = t.catch(() => {
    }), t;
  }
}();
await new Promise((r, e) => {
  const s = (t) => {
    const n = d.create(t, { type: "handshake" });
    n.ok && (window.removeEventListener("message", s), n.respond({ foo: 42 }), use.meta.server.targets = n.detail.server.targets, r(n.detail));
  };
  window.addEventListener("message", s);
});
use.sources.add("@@", async ({ options: r, owner: e, path: s }, ...t) => r.visible ? await l.show(s.specifier, r, ...t) : await l.request(s.specifier, r, ...t));
use.sources.add("rpc", async ({ options: r, owner: e, path: s }, ...t) => await l.request(s.specifier, r, ...t));
use.sources.add("api", async ({ options: r, owner: e, path: s }, ...t) => await l.request(s.specifier, r, ...t));
export {
  l as iworker
};
